text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,362 @@
+dependencies to all other crates in a repo. Additionally, we don't want to
+traverse wildly across the filesystem so we only go upwards to a fixed point or
+downwards to specific locations.
+
+### "Virtual" `Cargo.toml`
+
+A good number of projects do not have a root `Cargo.toml` at the top of a
+repository, however. While the explicit `package.workspace` and
+`workspace.members` keys should be enough to configure the workspace in addition
+to the implicit relations above, this directory structure is common enough that
+it shouldn't require *that* much more configuration.
+
+To accommodate this project layout, Cargo will now allow for "virtual manifest"
+files. These manifests will currently **only** contains the `[workspace]` key
+and will notably be lacking a `[project]` or `[package]` top level key.
+
+A virtual manifest does not itself define a crate, but can help when defining a
+root. For example a `Cargo.toml` file at the root of a repository with a
+`[workspace]` key plus `workspace.members` configuration would suffice for the
+project configurations in question. Note that omitting `workspace.members` would
+not be useful as there are no outgoing edges (no `path` dependencies), so Cargo
+will emit an error in cases like this.
+
+Cargo will for the time being disallow many commands against a virtual manifest,
+for example `cargo build` will be rejected. Arguments that take a package,
+however, such as `cargo test -p foo` will be allowed. Workspaces can eventually
+get extended with `--all` flags so in a workspace root you could execute
+`cargo build --all` to compile all crates.
+
+### Constructing a workspace
+
+With the explicit and implicit relations defined above, each crate will have a
+number of outgoing edges to other crates via `workspace.members`, path
+dependencies, and `package.workspace`. Two crates are then in the same workspace
+if they both transitively have edges to one another. A valid workspace then has
+exactly one root crate with a `[workspace]` key.
+
+While the restriction of one-root-per workspace may make sense, the restriction
+of crates transitively having edges to one another may seem a bit odd. The
+intention is to ensure that the set of packages in a workspace is the same
+regardless of which package is selected to start discovering a workspace from.
+
+With the implicit relations defined it's possible for a repository to not have a
+root package yet still have path dependencies. In this situation each dependency
+would not know how to get back to the "root package", so the workspace from the
+point of view of the path dependencies would be different than that of the root
+package. This could in turn lead to `Cargo.lock` getting out of sync.
+
+To alleviate misconfiguration, however, if the `workspace.members`
+configuration key contains a crate which is not a member of the constructed
+workspace, Cargo will emit an error indicating as such.
```

Comment:
> But workspace.members defines members of the workspace, how could mistakenly include packages?

That's what the configuration check is for, If A says B is in a workspace, that's only true if B says A is in its workspace as well. Authors could forget half of that relation, and Cargo will generate an informative error in that situation.

The paragraph you're quoting is indicating that when Cargo encounters that sort of relation via path dependencies it will not run into a bad situation where workspaces are different depending on what crate they're viewed from. The wording seems a little confusing, though, so I'll touch it up.